### PR TITLE
Added external V1 API functionality

### PR DESCRIPTION
**Pull request by kurtisassad**
Note: Some changes have been made to imports of files and to the subscription model. This is because loading them in causes tests relying on Sequelize or TokenBalanceCache to take a considerably long time to load. With these changes some tests have been sped up considerably.

## Description
Added endpoints for external API

## How has this been tested?
Added unit tests

## Does this PR affect any server routes?
- [x] yes
- [ ] no

## If this PR affects server routes, what are the security implications?
Added new server routes, we will probably need some rate-limiting mechanism if we decide to make these endpoints public.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

### CODE REVIEW

File: packages/commonwealth/server/models/subscription/subscription.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import Sequelize, { DataTypes} from 'sequelize';
```
Comment (jnaviask) [inline packages/commonwealth/server/models/subscription/subscription.ts:1]: Why was this moved into a folder?

File: packages/commonwealth/test/unit/unitHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { TypedRequestQuery, TypedResponse } from 'server/types';
```
Comment (jnaviask) [inline packages/commonwealth/test/unit/unitHelpers.ts:1]: Are these used anywhere?

File: packages/token-balance-cache/src/tbc.ts [hunk 3]
```diff
@@ -50,9 +49,11 @@ export class TokenBalanceCache extends JobRunner<ICache> implements ITokenBalanc
   constructor(
     noBalancePruneTimeS: number = 5 * 60,
     private readonly _hasBalancePruneTimeS: number = 1 * 60 * 60,
-    providers: BalanceProvider[] = BalanceProviders,
+    providers: BalanceProvider[] = null,
     private readonly _nodesProvider: (lastQueryUnixTime: number) => Promise<IChainNode[]> = queryChainNodesFromDB,
   ) {
+    // lazy load import to improve test speed
+    if(providers == null) import('./providers').then((p) => providers = p.default);
```
Comment (jnaviask) [inline packages/token-balance-cache/src/tbc.ts:56]: Race condition here with below code?

**Review by jnaviask — COMMENTED** [review]
Couple of general questions.

File: packages/commonwealth/server/models/subscription/subscription.ts (hunk 1 above)
Comment (kurtisassad) [reply packages/commonwealth/server/models/subscription/subscription.ts:1]: This needed to be moved out of the Subscriptions class, because it was causing tests relying on the db (which relied on the Subscription model) to take quite a long time to load. If we split this method from Subscription, only tests that rely on this emitNotification feature will take a long time to load.

File: packages/commonwealth/test/unit/unitHelpers.ts (hunk 2 above)
Comment (kurtisassad) [reply packages/commonwealth/test/unit/unitHelpers.ts:1]: There is a second issue with test speed. Basically there seem to be 2 ways to perform these tests.
1. load the whole app.ts, which gets all the endpoints. The problem with this way is it takes extremely long to load
2. call the methods that process the request by itself (ex, calling getComments instead of GET api/comments)

Since 1 takes too long to load due to it needing to basically build the whole project without webpack, I had created tested these endpoints using method 2. For this reason we need to use these helper methods in this class to build the express Request and Response to pass into the methods that process the request (ex `getComments(models:DB, req: TypedRequestQuery<GetCommentsReq>,res: TypedResponse<GetCommentsResp>)`.

Although thinking about it now, maybe testing it this way is not good, since there could be some issues that testing this way might not catch, what do you think?

File: packages/token-balance-cache/src/tbc.ts (hunk 3 above)
Comment (kurtisassad) [reply packages/token-balance-cache/src/tbc.ts:56]: Thanks for catching this

File: packages/commonwealth/server/models.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ import {
 } from './models/starred_community';
 import {
   SubscriptionModelStatic
-} from './models/subscription';
+} from 'server/models/subscription/subscription';
```
Comment (jnaviask) [inline packages/commonwealth/server/models.ts:79]: Can we use relative imports here?

File: packages/commonwealth/server/models/notifications_read.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import * as Sequelize from 'sequelize';
 import { Model, DataTypes } from 'sequelize';
 import { ModelStatic, ModelInstance } from './types';
-import { SubscriptionAttributes } from './subscription';
+import { SubscriptionAttributes } from 'server/models/subscription/subscription';
```
Comment (jnaviask) [inline packages/commonwealth/server/models/notifications_read.ts:4]: Can we use relative imports here?

File: packages/commonwealth/server/models/subscription/subscription.ts (hunk 1 above)
Comment (jnaviask) [reply packages/commonwealth/server/models/subscription/subscription.ts:1]: Now that emitNotification is in utils, we can move this back into the models folder?

File: packages/commonwealth/server/routes/createRole.ts [hunk 6]
```diff
@@ -5,7 +5,7 @@ import { DB } from '../models';
 import { success, TypedRequestBody, TypedResponse } from '../types';
 import { AppError } from '../util/errors';
 import { RoleAttributes } from '../models/role';
-import { SubscriptionAttributes } from '../models/subscription';
+import { SubscriptionAttributes } from 'server/models/subscription/subscription';
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/createRole.ts:8]: Can we use relative imports here?

File: packages/commonwealth/server/routes/getBalanceProviders.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import { DB } from 'server/models';
+import {
+  GetBalanceProvidersReq, GetBalanceProvidersResp,
+} from 'common-common/src/api/extApiTypes';
+import { TokenBalanceCache } from 'token-balance-cache/src';
+import { query, validationResult } from "express-validator";
+import { failure, success, TypedRequestQuery, TypedResponse } from '../types';
+
+export const getBalanceProvidersValidation = [
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/getBalanceProviders.ts:9]: Love this style.

File: packages/commonwealth/server/util/emitNotifications.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
+import { ChainBase, ChainType } from 'common-common/src/types';
```
Comment (jnaviask) [inline packages/commonwealth/server/util/emitNotifications.ts:1]: Can we use relative imports here?

File: packages/commonwealth/server/util/queries.ts [hunk 9]
```diff
@@ -1,64 +1,50 @@
-import { Sequelize } from 'sequelize';
-import { Models } from '../models';
-
+import { IPagination, OrderByOptions } from 'common-common/src/api/extApiTypes';
+import { failure } from 'server/types';
```
Comment (jnaviask) [inline packages/commonwealth/server/util/queries.ts:2]: Can we use relative imports here?

File: packages/commonwealth/test/unit/unitHelpers.ts (hunk 2 above)
Comment (jnaviask) [reply packages/commonwealth/test/unit/unitHelpers.ts:1]: I would call (1) an integration test and (2) a unit test. Both are good but unit tests are better.

**Review by zakhap — APPROVED** [review]

File: packages/commonwealth/server/routes/comments/getComments.ts [hunk 10]
```diff
@@ -1,27 +1,53 @@
...
+
+  let comments, count;
+  if (!count_only) {
+    ({ rows: comments, count } = await models.Comment.findAndCountAll({
+      where,
+      include,
+      ...formatPagination(req.query)
+    }));
+  } else {
+    count = <any>await models.Comment.count({
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/comments/getComments.ts:45]: Any reason for `any` cast here?

File: packages/commonwealth/server/routes/communities/getCommunities.ts [hunk 11]
```diff
@@ -1,25 +1,49 @@
-import { TypedRequestQuery, TypedResponse, success } from '../../types';
+import { GetCommunitiesReq, GetCommunitiesResp, needParamErrMsg } from 'common-common/src/api/extApiTypes';
+import { oneOf, query, validationResult } from "express-validator";
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/communities/getCommunities.ts:2]: Nitpick but can we be consistent with single quotes throughout this PR?

File: packages/commonwealth/server/routes/communities/getCommunities.ts [hunk 12]
```diff
@@ -1,25 +1,49 @@
...
 ) => {
-  return success(res, []);
+  const errors = validationResult(req).array();
+  if (errors.length !== 0) {
+    return failure(res.status(400), errors);
+  }
+  const { community_id, network, count_only } = req.query;
+
+  let where;
+  if(community_id) where = { id: community_id };
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/communities/getCommunities.ts:30]: Another nitpick, can we be consistent about spacing?

**Review by jnaviask — APPROVED** [review]
Couple small comments but nothing blocking!
